packages/rest/src/coercion/coerce-parameter.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  SchemaObject,
+  ReferenceObject,
+  isReferenceObject,
+} from '@loopback/openapi-v3-types';
+
+import * as HttpErrors from 'http-errors';
+
+/**
+ * Coerce the http raw data to a JavaScript type data of a parameter
+ * according to its OpenAPI schema specification.
+ *
+ * @param data The raw data get from http request
+ * @param schema The parameter's schema defined in OpenAPI specification
+ */
+export function coerceParameter(
+  data: string,
+  schema?: SchemaObject | ReferenceObject,
+) {
+  // ignore reference schema
+  if (!schema || isReferenceObject(schema)) return data;
+  let coercedResult;
+  coercedResult = data;
+  const OAIType = getOAIPrimitiveType(schema.type, schema.format);
+
+  // Review Note: [Validation place 1] Validation rules can be applied for each case
+  switch (OAIType) {
+    case 'byte':
+      coercedResult = Buffer.from(data, 'base64');
+      break;
+    case 'date':
+      coercedResult = new Date(data);
```

Comment:
How do we want to treat numbers (i.e. string containing only digits)?

Right now, you are treating them as the year:

```
> new Date('2014')
2014-01-01T00:00:00.000Z
```

Should we treat them as a timestamp instead? (Think of `Date.now()`) 

```
> new Date(2014)
1970-01-01T00:00:02.014Z
```

Make sure to write a test for this edge case!

Consider negative values too:

```
> new Date('-1')
2000-12-31T23:00:00.000Z
> new Date(-1)
1969-12-31T23:59:59.999Z
```

Also IIRC, OpenAPI Spec has two date-related types: `date` and `date-time`. We may want to enforce stricter rules for `dates` to ensure we don't accidentally include time portion in such values - a day starts at different time depending on the timezone.

Related reading: https://zachholman.com/talk/utc-is-enough-for-everyone-right

Reply:
same as https://github.com/strongloop/loopback-next/pull/1370#discussion_r192848933
I will track coming PRs in comment https://github.com/strongloop/loopback-next/issues/750#issuecomment-394481857

Reply:
@bajtos I think from the acceptance criteria we don't deal with milliseconds from Unix Epoch like `new Date('-1')` in this story. Neither of the OpenAPI date type(date and date-time) supports it. 
We may need additional type metadata provided from `@param`(or another new param decorator for coercion) to coerce/verify milliseconds.

Reply:
And by saying
> Right now, you are treating them as the year:
```
> new Date('2014')
2014-01-01T00:00:00.000Z
```
> Should we treat them as a timestamp instead? (Think of `Date.now()`) 
```
> new Date(2014)
1970-01-01T00:00:02.014Z
```
Could you elaborate more?

Reply:
>  I think from the acceptance criteria we don't deal with milliseconds from Unix Epoch like new Date('-1') in this story. Neither of the OpenAPI date type(date and date-time) supports it.

I see, that simplifies our life a lot! Can you please add test cases (and the corresponding implementation) to verify that our coercion algorithm rejects values that are not matching `full-date` or `date-time` as defined by [RFC3339](https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14)? See also [OpenAPI data types](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#data-types) which maps `date` and `date-time` OAI formats to RFC3339.

> Right now, you are treating them as the year:
> new Date('2014')
2014-01-01T00:00:00.000Z
> Should we treat them as a timestamp instead? (Think of Date.now())
> new Date(2014)
1970-01-01T00:00:02.014Z
Could you elaborate more?

Reply:
Nice now we are on the same page 👍 
> Can you please add test cases (and the corresponding implementation) to verify that our coercion algorithm rejects values that are not matching full-date or date-time as defined by RFC3339? See also OpenAPI data types which maps date and date-time OAI formats to RFC3339.

Definitely :) I was looking for some community validator to do that(that's why the datetime validation was in progress).